add query_by_tokens option in Neural Sparse Search (#7040)
opensearch-project/documentation-website: _query-dsl/specialized/neural-sparse.md

Introduced 2.11
{: .label .label-purple }

> Use the `neural_sparse` query for vector field search in [neural sparse search]({{site.url}}{{site.baseurl}}/search-plugins/neural-sparse-search/). Query can be performed using raw text or sparse vectors.

Reviewer: ```suggestion
Use the `neural_sparse` query for vector field search in [neural sparse search]({{site.url}}{{site.baseurl}}/search-plugins/neural-sparse-search/). The query can include both raw text and sparse vectors.
```

Author: The query should be used with either query text or tokens. Is it possible that the phrase `and` can cause confusion?

Reviewer: @zhichao-aws: I adjusted my comment, let me know if it works for you.

Reviewer 2: Works for me.

## Request fields

Include the following request fields in the `neural_sparse` query:

> 1. query by raw text:

Reviewer: ```suggestion
### Example: Query by raw text

```

  }
}
```

> 2. query by sparse vector:

Reviewer: ```suggestion
### Example: Query by sparse vector
```

```json
"neural_sparse": {
  "<vector_field>": {
    "query_tokens": "<query_tokens>"
  }
}
```

The top-level `vector_field` specifies the vector field against which to run a search query. The following table lists the other `neural_sparse` query fields.

Field | Data type | Required/Optional | Description
:--- | :--- | :--- 
`query_text` | String | Optional | The query text from which to generate sparse vector embeddings. 
`model_id` | String | Optional | The ID of the sparse encoding model or tokenizer model that will be used to generate vector embeddings from the query text. The model must be deployed in OpenSearch before it can be used in sparse neural search. For more information, see [Using custom models within OpenSearch]({{site.url}}{{site.baseurl}}/ml-commons-plugin/using-ml-models/) and [Neural sparse search]({{site.url}}{{site.baseurl}}/search-plugins/neural-sparse-search/). To set a default model id in neural sparse query, see [`neural_query_enricher`]({{site.url}}{{site.baseurl}}/search-plugins/search-pipelines/neural-query-enricher/).

> `query_tokens` | Map<String, Float> | Optional | The query tokens, also referred as sparse vector embeddings. Just like dense semantic retrieval, we can use raw sparse vectors generated by neural models or tokenizers to perform the semantic search. Either `query_text` or `query_tokens` must be provided for `neural_sparse` query.

Reviewer: ```suggestion
`query_tokens` | Map<String, Float> | Optional | The query tokens, sometimes referred to as sparse vector embeddings. Similar to dense semantic retrieval, use raw sparse vectors generated by neural models or tokenizers to perform a semantic search query. Use either the `query_text` option for raw field vectors, or `query_tokens` option for sparse vectors,  must be provided for the `neural_sparse` query to operate.
```

`max_token_score` | Float | Optional | (Deprecated) The theoretical upper bound of the score for all tokens in the vocabulary (required for performance optimization). For OpenSearch-provided [pretrained sparse embedding models]({{site.url}}{{site.baseurl}}/ml-commons-plugin/pretrained-models/#sparse-encoding-models), we recommend setting `max_token_score` to 2 for `amazon/neural-sparse/opensearch-neural-sparse-encoding-doc-v1` and to 3.5 for `amazon/neural-sparse/opensearch-neural-sparse-encoding-v1`. This field has been deprecated as of OpenSearch 2.12.

#### Example request

> 1. query by raw text

Reviewer: ```suggestion
**Query by raw text**

```

  }
}
```
### Example: Query by sparse vector
```json
"neural_sparse": {
  "<vector_field>": {
    "query_tokens": "<query_tokens>"
  }
}
```

The top-level `vector_field` specifies the vector field against which to run a search query. The following table lists the other `neural_sparse` query fields.

Field | Data type | Required/Optional | Description
:--- | :--- | :--- 
`query_text` | String | Optional | The query text from which to generate sparse vector embeddings. 

> `model_id` | String | Optional | The ID of the sparse encoding model or tokenizer model that will be used to generate vector embeddings from the query text. The model must be deployed in OpenSearch before it can be used in sparse neural search. For more information, see [Using custom models within OpenSearch]({{site.url}}{{site.baseurl}}/ml-commons-plugin/using-ml-models/) and [Neural sparse search]({{site.url}}{{site.baseurl}}/search-plugins/neural-sparse-search/). To set a default model id in neural sparse query, see [`neural_query_enricher`]({{site.url}}{{site.baseurl}}/search-plugins/search-pipelines/neural-query-enricher/).

Reviewer 2: ```suggestion
`model_id` | String | Optional | The ID of the sparse encoding model or tokenizer model that will be used to generate vector embeddings from the query text. The model must be deployed in OpenSearch before it can be used in sparse neural search. For more information, see [Using custom models within OpenSearch]({{site.url}}{{site.baseurl}}/ml-commons-plugin/using-ml-models/) and [Neural sparse search]({{site.url}}{{site.baseurl}}/search-plugins/neural-sparse-search/). For information on setting a default model ID in a neural sparse query, see [`neural_query_enricher`]({{site.url}}{{site.baseurl}}/search-plugins/search-pipelines/neural-query-enricher/).
```

> `model_id` | String | Optional | The ID of the sparse encoding model or tokenizer model that will be used to generate vector embeddings from the query text. The model must be deployed in OpenSearch before it can be used in sparse neural search. For more information, see [Using custom models within OpenSearch]({{site.url}}{{site.baseurl}}/ml-commons-plugin/using-ml-models/) and [Neural sparse search]({{site.url}}{{site.baseurl}}/search-plugins/neural-sparse-search/). To set a default model id in neural sparse query, see [`neural_query_enricher`]({{site.url}}{{site.baseurl}}/search-plugins/search-pipelines/neural-query-enricher/).

Reviewer 2: Confirm that my changes reflect the intended meaning.

> `query_tokens` | Map<String, Float> | Optional | The query tokens, sometimes referred to as sparse vector embeddings. Similar to dense semantic retrieval, use raw sparse vectors generated by neural models or tokenizers to perform a semantic search query. Use either the `query_text` option for raw field vectors, or `query_tokens` option for sparse vectors,  must be provided for the `neural_sparse` query to operate.

Reviewer 2: ```suggestion
`query_tokens` | Map<String, Float> | Optional | The query tokens, sometimes referred to as sparse vector embeddings. Similarly to dense semantic retrieval, you can use raw sparse vectors generated by neural models or tokenizers to perform a semantic search query. Use either the `query_text` option for raw field vectors or the `query_tokens` option for sparse vectors. Must be provided in order for the `neural_sparse` query to operate.
```

  }
}
```

> 2. query by sparse vector:

Reviewer: ```suggestion
**Query by sparse vector**

```
